File: 0048-rotate-image/0048-rotate-image.py
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        1. Transpose the matrix
        2. Reverse each row
        """

        self.transpose(matrix)
        for row in matrix:
            row.reverse()
        return
    
    def transpose(self, matrix: List[List[int]]) -> None:
        n_row = len(matrix)
        n_col = len(matrix[0])

        for r in range(n_row):
            for c in range(r+1, n_col):
                matrix[r][c], matrix[c][r] = matrix[c][r], matrix[r][c]
        return
        

File: 0048-rotate-image/0048-rotate-image.java
class Solution {
    public void rotate(int[][] matrix) {
        var n_row = matrix.length;
        var n_col = matrix[0].length;

        transpose(matrix);

        for (int r = 0; r < n_row; r++) {
            reverse(matrix[r]);
        }
    }

    private void reverse(int[] list) {
        var n = list.length;
        var l = 0;
        var r = n - 1;
        while (l < r) {
            var tmp = list[l];
            list[l] = list[r];
            list[r] = tmp;

            l++;
            r--;
        }
    }

    private void transpose(int[][] matrix) {
        var n_row = matrix.length;
        var n_col = matrix[0].length;

        for (int r = 0; r < n_row; r++) {
            for (int c = r + 1; c < n_col; c++) {
                var tmp = matrix[r][c];
                matrix[r][c] = matrix[c][r];
                matrix[c][r] = tmp;
            }
        }
    }
}